MPCA switch to svd solver

sz144 opened the pull request:
Fixes https://github.com/pykale/pykale/issues/61#issue-812515402.

### Description

- Solve eigen decomposition by `scipy.linalg.svd` instead of `numpy.linalg.eig`.
- Fix idx_order only sort top n_samples features.
- Update versions of dependencies in setup.py
- Minor changes in docstring

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] In-line docstrings updated and documentation `docs` updated.

haipinglu commented on kale/embed/mpca.py:
```diff
@@ -74,7 +74,7 @@ class MPCA(BaseEstimator, TransformerMixin):
 
     Attributes:
-        proj_mats (list of array-like): A list of transposed projection matrices, shapes (P_1, I_1), ...,
+        proj_mats (list of arraies): A list of transposed projection matrices, shapes (P_1, I_1), ...,
```
Typo: "arraies" should "arrays", see https://uk.mathworks.com/help/matlab/math/multidimensional-arrays.html

haipinglu commented on kale/embed/mpca.py:
```diff
@@ -153,7 +153,8 @@ def _fit(self, x):
         for i in range(1, n_dims):
-            eig_values, eig_vectors = np.linalg.eig(phi[i])
+            # eig_values, eig_vectors = np.linalg.eig(phi[i])
+            eig_vectors, eig_values, vt = np.linalg.svd(phi[i])
```
- SVD returns singular values rather than eigenvalues, so it is misleading here that you use variable `eig_values` for singular values (sqrt of eigenvalues).
- `phi[i]` should be `d x n` so I think it is correct to take the left singular vector but it will be good to verify it. I assume that you've did that, did you?

sz144 replied:
- Will change. Is `u` `s` `vt` better?
- `phi[i]` in shape `I_i x I_i`, it is a sysmetric positive definite matrix, so theoretically `u` = `vt.T` for those vectors corresponding to non-zero singular values. I have checked this in debuging and we can consider this as a test case

haipinglu replied:
@sz144 This is **terribly wrong** and it is clear that you did not verify the math or code.  You should really do some unit testing after making such changes. Study the slides posted, you should do SVD on the data matrix X, but now you are doing SVD on X^TX. You should get a piece of paper to work it out and make sure you understand before coding.

This makes me aware that SVD may not be efficient because the _data matrix_ here is much larger than the co-variance matrix so it may take more time. You may study the time cost of both.

sz144 replied:
In initial MATLAB implementation, we do eigen-decomposition for the **sum of covariance matrices**, but not the covariance of a certain matrix. So I do not know which matrix in `d x n` can be decomposed by SVD.

haipinglu replied:
That matrix is large. Considering mode-1: the matrix would be $I_1 \times (M \times I_2 \times \cdots \times I_N)$. And so we need to see how efficient SVD is performed on such a large matrix in numpy. We need to test it (for speed/space vs `eig`. If not, settling with `eig` may be better. `np.linalg.svd` does not allow to compute only `u` or only `vt`.

MPCA is doing X*X^T so getting the left singular vector is correct.

haipinglu replied:
The data matrix is effectively unfolding the whole (N+1)th-order data tensor along the nth mode.

sz144 replied:
I see, thank you. I have tried `scipy.linalg.svd` where make `full_matrices=False`, and it works pretty fast. Memory error will be raised if `full_matrices=True`. We then need to compare/verify the results obtained by SVD and EIG.

haipinglu replied:
Got it. That's good to know. Yes, `full_matrices=False` is correct and it is not meaningful otherwise. That is a relief and if we implement this way, the code will be more compact too.

haipinglu commented on kale/embed/mpca.py:
```diff
@@ -182,15 +183,16 @@ def _fit(self, x):
 
-                eig_values, eig_vectors = np.linalg.eig(phi[i])
+                # eig_values, eig_vectors = np.linalg.eig(phi[i])
+                eig_vectors, eig_values, vt = np.linalg.svd(phi[i])
```
The same here.

haipinglu commented on kale/embed/mpca.py:
```diff
@@ -182,15 +183,16 @@ def _fit(self, x):
         x_proj_unfold = unfold(x_projected, mode=0)  # unfold the tensor projection to shape (n_samples, n_features)
         # x_proj_cov = np.diag(np.dot(x_proj_unfold.T, x_proj_unfold))  # covariance of unfolded features
-        x_proj_cov = np.sum(np.multiply(x_proj_unfold, x_proj_unfold), axis=1)  # memory saving computing covariance
+        x_proj_cov = np.sum(np.multiply(x_proj_unfold.T, x_proj_unfold.T), axis=1)  # memory saving computing covariance
```
`.T` means the transpose, right? Is it wrong previously? We should have some testing implemented to verify the correctness of code. I understand that we are at level zero but MPCA maybe a good starting point since I/we know best how it should behave.

sz144 replied:
It was wrong. Previously we take the diagnoal of covariance matrix (commented above). It was changed to `np.multiply` in the last update because Cameron reported a memory issue of `np.dot`. Then Lawrence pointed this out and I corrected in this PR. I agree to have a test for mpca. I will implement a demo test.

haipinglu replied:
This type of problem is identifiable automatically through testing so I do think it is time for us to consider it.

haipinglu reviewed (changes requested):
Let us confirm the changes are correct and fix other minor issues before merging. Thanks.

haipinglu commented on kale/embed/mpca.py:
```diff
@@ -14,6 +14,7 @@
 import warnings
 
 import numpy as np
+from scipy import linalg
```
So is `scipy.linalg` better than `np.linalg`? If yes, `scipy` needs to be added to `setup.py` `install..._req..."

haipinglu commented on kale/embed/mpca.py:
```diff
@@ -144,52 +144,43 @@ def _fit(self, x):
-                if i not in phi:
-                    phi[i] = 0
-                phi[i] = phi[i] + np.dot(x_ji, x_ji.T)
+            phi[i] = unfold(x, mode=i)
```
Consider a new variable name. This is not longer the same as in the Matlab code or paper so a more descriptive name would be better, e.g. `mode_data_mat`. Moreover, why creating these for all modes and then do the SVD (increasing the memory requirement nx? This for loop should be combined with the for loop below to be more space (and time) efficient.

haipinglu commented on kale/embed/mpca.py:
```diff
@@ -144,52 +144,43 @@ def _fit(self, x):
+            u, s, v = linalg.svd(phi[i], full_matrices=False)
+            idx_sorted = (-1 * s).argsort()
+            cum = s[idx_sorted]
             var_tot = np.sum(cum)
```
`cum` here is not the variance. Singular values are not equal to eigenvalues.

haipinglu commented on kale/embed/mpca.py:
```diff
@@ -144,52 +144,43 @@ def _fit(self, x):
                     shape_out += (j,)
                     break
-            proj_mats.append(eig_vectors[:, idx_sorted][:, : shape_out[i - 1]].T)
+            proj_mats.append(u[:, idx_sorted][:, : shape_out[i - 1]].T)
```
u is not as readable

sz144 replied:
It seems `u`, `s` and `v` are common variable names for SVD in popular python packages like scipy, numpy, scikit-learn, pytorch, and tensorly. That is why `u` is used here. To improve readability, `eig_vectors` or `singular_vectors_left` probably are good choices.

haipinglu replied:
I think so, not all know the relationship between svd and eigendecomposition.

haipinglu commented on kale/embed/mpca.py:
```diff
@@ -144,53 +144,46 @@ def _fit(self, x):
+                    [proj_mats[m] for m in range(n_dims - 1) if m != i - 1],
+                    modes=[m for m in range(1, n_dims) if m != i]
+                )
+                phi[i] = unfold(x_projected, i)
```
The same comment as the above as the name (though no memory space problem here)

haipinglu reviewed (commented):
If unclear, we can discuss when we meet or schedule additional meetings if needed.
I shared a recorded meeting with you. Since MPCA has been buggy, it will be the best to design test cases based on Matlab code (as truth) to make sure that the code is working as we want. We could work together if needed. The [testing reference from Bob](https://carpentries-incubator.github.io/python-testing/04-units/index.html) is really good (I have studied it and now working on writing test cases).

We could consider using the Gallery gait data 32x22x10 as the data for testing MPCA, and use PrbA for classification pipeline, both are only 278KB in mat.

sz144 commented:
> 
> 
> If unclear, we can discuss when we meet or schedule additional meetings if needed.
> I shared a recorded meeting with you. Since MPCA has been buggy, it will be the best to design test cases based on Matlab code (as truth) to make sure that the code is working as we want. We could work together if needed. The [testing reference from Bob](https://carpentries-incubator.github.io/python-testing/04-units/index.html) is really good (I have studied it and now working on writing test cases).
> 
> We could consider using the Gallery gait data 32x22x10 as the data for testing MPCA, and use PrbA for classification pipeline, both are only 278KB in mat.

It is great, shall we upload the gait data to the repo under pykale/tests/test_data/? I have implemented a demo (https://github.com/sz144/pykale/blob/mpca_test/tests/embed/test_mpca.py). I think it need to be seperated to several individual test functions and more test cases maybe needed. We can discuss in our next meeting.

haipinglu commented:
> > If unclear, we can discuss when we meet or schedule additional meetings if needed.
> > I shared a recorded meeting with you. Since MPCA has been buggy, it will be the best to design test cases based on Matlab code (as truth) to make sure that the code is working as we want. We could work together if needed. The [testing reference from Bob](https://carpentries-incubator.github.io/python-testing/04-units/index.html) is really good (I have studied it and now working on writing test cases).
> > We could consider using the Gallery gait data 32x22x10 as the data for testing MPCA, and use PrbA for classification pipeline, both are only 278KB in mat.
> 
> It is great, shall we upload the gait data to the repo under pykale/tests/test_data/? I have implemented a demo (https://github.com/sz144/pykale/blob/mpca_test/tests/embed/test_mpca.py). I think it need to be seperated to several individual test functions and more test cases maybe needed. We can discuss in our next meeting.

- Yes, I think we could upload it, either in `mat` or `pickle`. It is small and has uneven dimension 32x22x10, good for testing.
- Do we have mat conversion implemented?
- I think you should keep your `test_mpca` and `mpca` under one branch so you could actually test it more conveniently. If needed, you could get help from Bob regarding testing. He previously helped converting GPy Matlab to Python.
- I added one card @ you suggesting that _Our MPCA-Fisher-SVM/LR pipeline may be implemented as a pipeline in kale.pipeline since it is highly reusable_.

bobturneruk commented:
Maybe this is useful as an example of testing python against a matlab baseline: https://github.com/RSE-Sheffield/GPy/blob/lfm-devel/GPy/testing/lfmXlfm_unit_tests.py

sz144 commented:
> 
> 
> > > If unclear, we can discuss when we meet or schedule additional meetings if needed.
> > > I shared a recorded meeting with you. Since MPCA has been buggy, it will be the best to design test cases based on Matlab code (as truth) to make sure that the code is working as we want. We could work together if needed. The [testing reference from Bob](https://carpentries-incubator.github.io/python-testing/04-units/index.html) is really good (I have studied it and now working on writing test cases).
> > > We could consider using the Gallery gait data 32x22x10 as the data for testing MPCA, and use PrbA for classification pipeline, both are only 278KB in mat.
> > 
> > 
> > It is great, shall we upload the gait data to the repo under pykale/tests/test_data/? I have implemented a demo (https://github.com/sz144/pykale/blob/mpca_test/tests/embed/test_mpca.py). I think it need to be seperated to several individual test functions and more test cases maybe needed. We can discuss in our next meeting.
> 
>     * Yes, I think we could upload it, either in `mat` or `pickle`. It is small and has uneven dimension 32x22x10, good for testing.
> 
>     * Do we have mat conversion implemented?
> 
>     * I think you should keep your `test_mpca` and `mpca` under one branch so you could actually test it more conveniently. If needed, you could get help from Bob regarding testing. He previously helped converting GPy Matlab to Python.
> 
>     * I added one card @ you suggesting that _Our MPCA-Fisher-SVM/LR pipeline may be implemented as a pipeline in kale.pipeline since it is highly reusable_.

I will upload .mat data and merge `test_mpca` into `mpca` after merging this PR.

haipinglu commented on setup.py:
```diff
@@ -12,10 +12,11 @@
 # python -m pip install torch-cluster torch-scatter torch-sparse torch-spline
 requirements = [
     'numpy>=1.18.0',
+    'scipy>=1.5.4',
```
This list is sorted so you should put it after `scikit` and before `t`

haipinglu commented:
@sz144 The description needs to be updated since you are using scipy for SVD now.

haipinglu commented:
@sz144 In Spark implementation, they set all eigenvalues smaller than 10^{-9) x the largest eigenvalue to zero by default.
